Add all_paths utility function in order to show why modules are included in dependencies

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65602

This pr allows users of PackageExporter to know why a certain module has been included in the packaged dependencies. It shows all dependencies from `src` to `dst`.

fixes: https://github.com/pytorch/pytorch/issues/65261
Differential Revision: [D31163681](https://our.internmc.facebook.com/intern/diff/D31163681)

tugsbayasgalan committed:
Add all_paths utility function

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31163681).

tugsbayasgalan committed:
Update on "Add all_paths utility function"


fixes: https://github.com/pytorch/pytorch/issues/65261
Differential Revision: [D31163681](https://our.internmc.facebook.com/intern/diff/D31163681)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31163681).

suo commented on torch/package/package_exporter.py:
```diff
@@ -1032,6 +1032,15 @@ def get_rdeps(self, module_name: str) -> List[str]:
+        """Return a subgraph that has all paths from src to dst.
+
+        Returns:
+            A DiGraph containing all paths from src to dst.
```
I changed my mind about this return value. One of the issues is that `DiGraph` is not a public API and is typically in a format that doesn't make sense to external users. Can we use the same functionality that is in `dependency_graph_string` to return a graphviz string instead of a raw `DiGraph`?

tugsbayasgalan replied:
done

suo commented on torch/package/_digraph.py:
```diff
@@ -83,3 +86,53 @@ def __contains__(self, n):
         except TypeError:
             return False
+
+    def forward_transitive_closures(self, src: str) -> Set[str]:
```
nit: shouldn't be plural, so `forward_transitive_closure`

tugsbayasgalan replied:
done

suo commented on torch/package/_digraph.py:
```diff
@@ -83,3 +86,53 @@ def __contains__(self, n):
+                    working_set.append(n)
+        return result
+
+    def backward_transitive_closures(self, src: str) -> Set[str]:
```
same as above

tugsbayasgalan replied:
done

suo reviewed (changes requested):
Excellent work, only minor comments!

tugsbayasgalan committed:
Update on "Add all_paths utility function"


fixes: https://github.com/pytorch/pytorch/issues/65261
Differential Revision: [D31163681](https://our.internmc.facebook.com/intern/diff/D31163681)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31163681).

tugsbayasgalan committed:
Update on "Add all_paths utility function"


fixes: https://github.com/pytorch/pytorch/issues/65261
Differential Revision: [D31163681](https://our.internmc.facebook.com/intern/diff/D31163681)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31163681).

suo reviewed (approved):
💯
